src/client.rs

```diff
@@ -115,19 +117,91 @@ pub fn connect<'t, Req: Into<Request<'t>>>(request: Req)
     connect_with_config(request, None)
 }
 
-fn connect_to_some<A>(addrs: A, url: &Url, mode: Mode) -> Result<AutoStream>
-    where A: Iterator<Item=SocketAddr>
-{
-    let domain = url.host_str().ok_or_else(|| Error::Url("No host name in the URL".into()))?;
+fn connect_to_some(req: &Request, mode: Mode) -> Result<AutoStream> {
+    let addrs = if let Some(ref p) = req.proxy {
+        p.url.to_socket_addrs()?
+    } else {
+        req.url.to_socket_addrs()?
+    };
+
+    let domain = req.url.host_str().ok_or_else(|| Error::Url("No host name in the URL".into()))?;
+
     for addr in addrs {
-        debug!("Trying to contact {} at {}...", url, addr);
-        if let Ok(raw_stream) = TcpStream::connect(addr) {
+        debug!("Trying to contact {} at {}...", req.url, addr);
+
+        if let Ok(mut raw_stream) = TcpStream::connect(addr) {
+            if let Some(ref proxy) = req.proxy {
+                let port = req.url.port_or_known_default().unwrap_or_else(|| {
+                    match mode {
+                        Mode::Plain => 80,
+                        Mode::Tls => 443
+                    }
+                });
+
+                // Connect to a proxy here
+                let mut buf = format!("\
+                    CONNECT {host}:{port} HTTP/1.1\r\n\
+                    Host: {host}:{port}\r\n\
+                    Proxy-Connection: Keep-Alive\r\n",
+                    host = domain,
+                    port = port,
+                ).into_bytes();
+
+                // Add any headers required for authentication
+                if let Some(ref auth) = proxy.auth {
+                    match auth {
+                        ProxyAuth::Basic(basic_auth) => {
+                            buf.extend_from_slice(&format!("Proxy-Authorization: {}\r\n", basic_auth.to_header_value()).as_bytes());
+                        }
+                    }
+                }
+
+                // Add the trailing empty line
+                buf.extend_from_slice(b"\r\n");
+
+                let _ = raw_stream.write(&buf)?;
```

Comment:
I might need to continue in the `for addr in addrs` loop instead of early -returning here.